TailLight/driver.cpp

```diff
@@ -11,13 +11,16 @@ NTSTATUS DriverEntry(
 
     WDF_DRIVER_CONFIG params = {};
     WDF_DRIVER_CONFIG_INIT(/*out*/&params, EvtDriverDeviceAdd);
-    params.DriverPoolTag = POOL_TAG;
     params.EvtDriverUnload = EvtDriverUnload;
 
+    WDF_OBJECT_ATTRIBUTES attributes = {};
+    WDF_OBJECT_ATTRIBUTES_INIT(&attributes);
+    WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&attributes, DRIVER_CONTEXT);
```

Comment:
Can you please explain the intent of setting "DRIVER_CONTEXT" here?
Also, I think the `WDF_OBJECT_ATTRIBUTES_INIT` call becomes redundant if you're afterwards calling `WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE`.

Reply:
The callback registered by IoRegisterPlugPlayNotification has to be unregistered when the driver unloads. Otherwise, the driver is called with nothing guaranteed to be present in handling any requests. EvtDriverUnload taked a driver object. Therefore the handle that controls the PlugPlayNotification registration is best stored in the driver objects's context space.

Reply:
Thanks for explaining.